### ISSUE TITLE
[Feature Request] gather non-tensor objects

### ISSUE THREAD

**Issue by Hannibal046** [huggingface/accelerate#1383]
Considering that PyTorch has already  support non-tensor objects. Can we just make accelerate to support this? This is so important to calculate metrics for NLP tasks since we don't want to decoder(encode(labels)).

https://pytorch.org/docs/stable/_modules/torch/distributed/distributed_c10d.html#all_gather_object

relevant issue: https://github.com/huggingface/accelerate/issues/476

**Comment by muellerzr** [comment]
This already exists, please use `accelerate.utils.gather_objects`

**Comment by Hannibal046** [comment]
Oh, i just found it. Thanks ! 

BTW, would you mind checking this issue https://github.com/huggingface/accelerate/issues/1382, it confuse me for a while.

**Comment by Hannibal046** [comment]
Hi, there seems to have some bugs about `gather_object`

For tensors we gather:
```python
## CUDA_VISIBLE_DEVICES=0,1 accelerate run --num_processes 2 test.py
import torch
from accelerate import Accelerator

accelerator = Accelerator()
dataloader = torch.utils.data.DataLoader(range(9), batch_size=5)
dataloader = accelerator.prepare(dataloader)
batch = next(iter(dataloader))
gathered_items = accelerator.gather_for_metrics(batch)
if accelerator.is_local_main_process:
    print(gathered_items)
```
This would output:
```
tensor([0, 1, 2, 3, 4, 5, 6, 7, 8], device='cuda:0')
```

For objects we gather:
```python
## CUDA_VISIBLE_DEVICES=0,1 accelerate run --num_processes 2 test.py
import torch
from accelerate import Accelerator
from accelerate.utils import gather_object

accelerator = Accelerator()
obj = [accelerator.local_process_index]
obj = gather_object(obj)
if accelerator.is_local_main_process:
    print(obj)
```
this would only return:
```
[0]
```

I using `torch.distributed` to rewrite this, and it works fine:
```python
from accelerate import Accelerator
import torch.distributed as dist

accelerator = Accelerator()
obj = [accelerator.local_process_index]
all_rank_outputs = [None for _ in range(accelerator.num_processes)]    
dist.all_gather_object(all_rank_outputs,obj) ## would create new proce
obj = [x for y in all_rank_outputs for x in y]
if accelerator.is_local_main_process:
    print(obj)
```
returns 
```
[0,1]
```

**Comment by muellerzr** [comment]
For now use `torch.dist.all_gather_object`, and I'll take a look at what could be going wrong with the current implementation. Thanks!

**Comment by muellerzr** [comment]
Thanks @Hannibal046, PR #1391 should fix this right up :)

**Comment by ywangmy** [comment]
Hi, it seems that `all_gather_object` does not remove duplicates for the last non-full batch, which is inconsistent with `gather_for_metrics` during evaluation.

**Comment by muellerzr** [comment]
Yes, only `gather_for_metrics` does the duplicate removal. If you see this with `gather_for_metrics`, please provide us a reproducer.

**Comment by ywangmy** [comment]
`gather_for_metrics` works fine for me. Is there any alternative function for non-tensors that removes duplicates when gathering?

Also, setting `drop_last=True` for the evaluation dataloader does not resolve the problem (since the batchsize changed?)

**Comment by muellerzr** [comment]
There is not. These are raw gather ops that are 1:1 with torch distributed. You'll find that `gather()` will also not de-dupe. You'll need to pay attention to this, or create a reproducer if `gather_for_metrics` not working and open a new feature request asking for us to enable it

**Comment by ywangmy** [comment]
Thanks for the information. I've created a new feature request #1925.
